### ISSUE TITLE
Add multiquery support to HTTP API

### ISSUE THREAD

**Issue by maxehmookau** [ClickHouse/ClickHouse#61608]
**Use case**

Currently the native connection has multiquery support (`SELECT 1; SELECT 2`) when using the `--multiquery` switch on the CLI. This works as expected when doing multiple queries in the same statement.

This support does not exist in the HTTP API and we (GitLab) would find it very useful. 

This has been raised in an issue _before_ multiquery support was added to the native client but was rejected. Now that this feature does exist in the native client, I would expect the HTTP client to have support too.

- Previous issue: https://github.com/ClickHouse/ClickHouse/issues/6707
- Previous, rejected, PR: https://github.com/ClickHouse/ClickHouse/pull/6710

**Describe the solution you'd like**

When making multiple queries in the same statement, I want ClickHouse server to respond in the same way regardless of whether using the native CLI or HTTP API. (`SELECT 1; SELECT 2` should work regardless of the API used - or atleast, we should have the option to enable it)

**Describe alternatives you've considered**

* Using the native connection, however our integrations are all using the HTTP client. 

**Additional context**

We're willing to do some or all of the development work if this is a change that is likely to be accepted.

**Comment by maxehmookau** [comment]
@jaijhala Pinging you on this as requested on CH support ticket `00009461`.

**Comment by den-crane** [comment]
>Currently the native connection has multiquery support (SELECT 1; SELECT 2)

It's not Native. It's a Clickhouse-client feature. Native protocol does not support it.

**Comment by maxehmookau** [comment]
> It's not Native. It's a Clickhouse-client feature. Native protocol does not support it.

Ok, so what is clickhouse-client doing that means its possible? It's the server that responds with an error, but the clickhouse-client is able to circumvent the server error? I'm not clear on what's happening there.

**Comment by den-crane** [comment]
Yes, 

```
 --ignore-error                                                                       do not stop processing in multiquery mode
```

ignore-error  is also an inside feature of Clickhouse-client

**Comment by maxehmookau** [comment]
So clickhouse-client can force the server to allow multiquery statements, and also ignore all server errors to allow the use of multiple queries, correct?

If so, is it possible to add the same opt-in for queries made via the HTTP API? Perhaps by the use of a header? `CH-MULTIQUERY=true` or something similar?

**Comment by den-crane** [comment]
clickhouse-client parses SQL script and executes SQL queries one by one.

**Comment by alexey-milovidov** [comment]
@maxehmookau, There are two problems to solve when implementing this feature.

1. Correct splitting of queries in the case of mixed INSERT queries with data and SELECT queries. This problem also exists in clickhouse-client. The proposal is here: https://github.com/ClickHouse/ClickHouse/issues/58436 If you are considering using the multi-query feature without INSERT queries, we can ignore this item.

2. Framing of responses from different queries: what is the format of outputting multiple resultsets? The simplest solution could be - no framing; the results are output one after another with no additional separators.

Let me know if we can do these shortcuts. Then, the implementation will be trivial.

**Comment by canhld94** [comment]
@maxehmookau can I know which use case requires this feature in GitLab?
FTR, at Ahrefs, we also use HTTP for 99% of queries. Back then, some developers have a subquery that the result will be used repeatedly in the query, so they need to split the original query into 3 queries: create the temporary table from subquery, run the main query, then drop the temporary table. These needs to be done in 3 different HTTP calls. So I was also requested to implement supporting multiple queries in one HTTP calls. However, after some attempts, I feel that it's hard to make this feature works correctly without pitfalls (as @alexey-milovidov point out).
So we took another approach: implement materialized CTE (https://github.com/ClickHouse/ClickHouse/pull/61086).
If you have the same use case, then our PR can also help you.

**Comment by maxehmookau** [comment]
> Let me know if we can do these shortcuts. Then, the implementation will be trivial.

@alexey-milovidov Unfortunately in our use-case, we would require both `CREATE TABLE` and `INSERT`.

@canhld94 I'll give the wider context if its helpful:

We're using [Cube](https://cube.dev/) as a headless BI tool and ClickHouse as the storage solution for it. Cube supports the ability to pre-aggregate data in to the same storage (CH for us). **To populate the pre-aggregated data, Cube uses a `CREATE AS SELECT` statement _which is not supported by ClickHouse Cloud_.** (https://clickhouse.com/docs/en/whats-new/cloud-compatibility#ddl-syntax) - We only discovered this incompatibility after we moved from self-hosted to cloud managed. (If we can fix that for Cloud users, then this entire issue isn't neccessary.)

The alternative is to use `CREATE ... EMPTY ... AS SELECT` followed by `INSERT` statement. Without an enormous refactor of the Cube codebase (which we tried, but they're not really interested in what we're trying to do as they're pushing folks towards their own storage system and the ClickHouse driver is only community-supported) our only option is to make a smaller change to the Cube codebase which sends both the `CREATE` and `INSERT` statements together at the same time.

So I guess this also leads me to ask _why isn't `CREATE AS SELECT` supported by CH Cloud? Is that on the roadmap? If so, that will save us a lot of time.

**Comment by tavplubix** [comment]
> Is that on the roadmap?

No, but we have another item in the roadmap, and it will replace `Replicated` databases in ClickHouse Cloud and will support CREATE SELECT. It's being actively developed right now and is expected to become available in a few months.

**Comment by maxehmookau** [comment]
@tavplubix That would solve our issue. However, presumably there's a tradeoff there for replacing Replicated databases? Can you share any more information about it?

**Comment by den-crane** [comment]
@alexey-milovidov 

I see it as it as sPL (procedure blocks) which can return only single result (single error / single dataset).

```
BEGIN
  set VAR1 = 'foo';
  create temporary table TEMP as .... ;
  insert into OTHER ...
  insert into TEMP ... 
  ....
  select .... -- result of the last select will be returned to the user (or if more than one select is found then it can throw an error)
END
```

If an error happened at any step the execution stops and returns the error and the index of the command or line number...

**Comment by maxehmookau** [comment]
> If an error happened at any step the execution stops and returns the error and the index of the command or line number...

This would be sufficient for us. There's no need for the multiple steps to be wrapped in a transaction. It's just the ability to send multiple queries to the server at once via HTTP that is important to us right now.

That said, I understand that this would be a fairly major change, so I'm keen to understand more about the new feature that would remove the need to use replicated tables if anything can be shared.

**Comment by brian-mann** [comment]
Another option here would be to support materialized CTE's, which would solve for the need to first create a temporary table and then query it, and would avoid the need to issue multiple queries. 

I opened a similar issue here https://github.com/ClickHouse/clickhouse-js/issues/283 for the exact same use case (creating a temporary table and then querying it immediately) and noticed there was a draft for materialized CTE's.
